src/Symfony/Component/Serializer/Mapping/Factory/CompiledClassMetadataFactory.php

```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Serializer\Mapping\Factory;
+
+/**
+ * @author Fabien Bourigault <bourigaultfabien@gmail.com>
+ */
+final class CompiledClassMetadataFactory implements ClassMetadataFactoryInterface
+{
+    private $compiledClassMetadata = array();
+
+    /**
+     * @var ClassMetadataFactoryInterface
+     */
+    private $classMetadataFactory;
+
+    public function __construct(array $compiledClassMetadata, ClassMetadataFactoryInterface $classMetadataFactory)
+    {
+        $this->compiledClassMetadata = $compiledClassMetadata;
+        $this->classMetadataFactory = $classMetadataFactory;
+    }
+
+    public function getMetadataFor($value)
+    {
+        $className = \is_object($value) ? \get_class($value) : $value;
+
+        if (isset($this->compiledClassMetadata[$className])) {
+            return unserialize($this->compiledClassMetadata[$className]);
```

Comment:
if we ask multiple time for the metadata for the same class, we should return the same instance (as done by the normal ClassMetadataFactory), not a new instance each time.